plugins/events-backend/src/service/hub/DatabaseEventBusStore.ts

```diff
@@ -0,0 +1,555 @@
+          )
+          .where('event.id', '>', this.#db.ref('read_until').withSchema('sub'))
+          .orderBy('event.id', 'asc')
+          .limit(MAX_BATCH_SIZE),
+      )
+      // Find the ID of the last event in the batch, for use as the new read_until marker
+      .with('last_event_id', q => q.max({ last_event_id: 'id' }).from('events'))
+      // Aggregate the events into a JSON array so that we can return all of them with the UPDATE
+      .with('events_array', q =>
+        q
+          .select({ events: this.#db.raw('json_agg(row_to_json(events))') })
+          .from('events'),
+      )
+      // Update the read_until marker to the ID of the last event, or if no
+      // events where read, the last ID out of all events
+      .update({
+        read_until: this.#db.raw(
+          'COALESCE(last_event_id, (SELECT MAX(id) FROM event_bus_events), 0)',
+        ),
+      })
+      .updateFrom({
+        events_array: 'events_array',
+        last_event_id: 'last_event_id',
+      })
+      .where(`${TABLE_SUBSCRIPTIONS}.id`, id)
+      .returning<[{ events: EventsRow[] }]>('events_array.events');
+
+    if (result.length === 0) {
+      throw new NotFoundError(`Subscription with ID '${id}' not found`);
+    } else if (result.length > 1) {
+      throw new Error(
+        `Failed to read subscription, unexpectedly updated ${result.length} rows`,
+      );
+    }
+
+    const rows = result[0].events;
+    if (!rows || rows.length === 0) {
+      return { events: [] };
+    }
+
+    return {
+      events: result[0].events.map(row => {
+        const { payload, metadata } = JSON.parse(row.data_json);
+        return {
+          topic: row.topic,
+          eventPayload: payload,
+          metadata,
+        };
+      }),
+    };
+  }
+
+  async setupListener(
+    subscriptionId: string,
+    options: {
+      signal: AbortSignal;
+    },
+  ): Promise<{ waitForUpdate(): Promise<{ topic: string }> }> {
+    const result = await this.#db<SubscriptionsRow>(TABLE_SUBSCRIPTIONS)
+      .select('topics')
+      .where({ id: subscriptionId })
+      .first();
+
+    if (!result) {
+      throw new NotFoundError(
+        `Subscription with ID '${subscriptionId}' not found`,
+      );
+    }
+
+    options.signal.throwIfAborted();
+
+    return this.#listener.setupListener(
+      new Set(result.topics ?? []),
+      options.signal,
+    );
+  }
+
+  #cleanup = async () => {
+    try {
+      const eventCount = await this.#db
+        .delete()
+        .from(TABLE_EVENTS)
+        // Delete any events that are outside both the min age and size window
+        .where('created_at', '<', new Date(Date.now() - this.#windowMinAge))
+        .andWhere(
+          'id',
+          '=',
+          this.#db
+            .raw(
+              this.#db
+                .select('id')
+                .from(TABLE_EVENTS)
+                .orderBy('id', 'desc')
+                .offset(this.#windowSize),
```

Comment:
Applying this as a sub-query might lead to more events being kept than defined by `windowSize` due to the previous filter. Maybe, this was done intentionally.

Otherwise, we could do two deletes in sequence.

1. Time-based deletion
2. Count-based deletion